### PR TITLE
Handle reused payment destinations without failing invoice creation

### PR DESCRIPTION
**Pull request by CaMoPeZzz**
When a payment method or plugin returns a destination that is already assigned to another invoice, invoice creation currently fails with a database constraint error and an HTTP 500 response. This is especially visible with Monero after wallet upgrades that may return a previously used address.

This change detects that conflict and disables only the affected payment method. The invoice can still be created when another payment method is available. Destination ownership is reserved atomically, so concurrent invoice creation and lazy payment-method activation cannot save a prompt that belongs to another invoice.

Lightning prompt rotation follows the same ownership rules and keeps the previous invoice usable until the replacement has been saved successfully.

Related issue: https://github.com/btcpay-monero/btcpayserver-monero-plugin/issues/57

Testing: Added coverage for sequential and concurrent destination conflicts, transaction rollback, lazy prompt activation, missing invoices, and atomic prompt updates. All 12 `DatabaseTests` pass against the current `master`.

No visual changes.

### CODE REVIEW

File: BTCPayServer/Controllers/UIInvoiceController.cs [hunk 1]
```diff
@@ -283,7 +319,22 @@ internal async Task<InvoiceEntity> CreateInvoiceCoreRaw(InvoiceEntity entity, St
...
+                        break;
+                    }
+                    catch (PaymentDestinationAlreadyTrackedException ex) when (creationContext.PaymentMethodContexts.TryGetValue(ex.PaymentMethodId, out var paymentContext))
+                    {
+                        DisablePaymentMethod(paymentContext, ex.InvoiceId);
+                        var contexts = GetActiveContexts();
+                        if (contexts.Count == 0)
+                            throw NoPaymentMethodException();
+                        entity.SetPaymentPrompts(new PaymentPromptDictionary(contexts.Select(c => c.Prompt)));
+                    }
```
[the remark this answers is not in the document]
Comment (CaMoPeZzz) [reply BTCPayServer/Controllers/UIInvoiceController.cs]: Fixed in 84389baea. The catch now handles every PaymentDestinationAlreadyTrackedException and converts an unknown payment method into the established 400 response.

File: BTCPayServer/Services/Invoices/InvoiceRepository.cs [hunk 2]
```diff
@@ -77,16 +93,54 @@ public async Task<InvoiceEntity> GetInvoiceFromAddress(PaymentMethodId paymentMe
...
+                .Select(a => a.InvoiceDataId)
+                .FirstOrDefaultAsync();
+        }
+
         public async Task AddAddressInvoice(string invoiceId, PaymentMethodId paymentMethodId, string address)
         {
             await using var context = _applicationDbContextFactory.CreateContext();
-            await UpsertAddressInvoice(context, invoiceId, paymentMethodId.ToString(), address);
+            await TryAddAddressInvoice(context, invoiceId, paymentMethodId.ToString(), address);
         }
```
[the remark this answers is not in the document]
Comment (CaMoPeZzz) [reply BTCPayServer/Services/Invoices/InvoiceRepository.cs]: Fixed in 84389baea. The public AddAddressInvoice path now uses the same owner validation and controlled conflict exception.

File: BTCPayServer/Services/InvoiceActivator.cs [hunk 3]
```diff
@@ -59,7 +59,8 @@ public async Task<bool> ActivateInvoicePaymentMethod(string invoiceId, PaymentMe
                     await paymentContext.CreatePaymentPrompt();
                     if (paymentContext.Status == PaymentMethodContext.ContextStatus.Created)
                     {
-                        await _invoiceRepository.NewPaymentPrompt(invoice.Id, paymentContext);
+                        if (!await _invoiceRepository.NewPaymentPrompt(invoice.Id, paymentContext))
+                            return false;
```
[the remark this answers is not in the document]
Comment (CaMoPeZzz) [reply BTCPayServer/Services/InvoiceActivator.cs]: Fixed in 84389baea. A false result now records an activation failure and reaches AddInvoiceLogs before returning.

File: BTCPayServer/Services/Invoices/InvoiceRepository.cs [hunk 4]
```diff
@@ -77,16 +93,54 @@ public async Task<InvoiceEntity> GetInvoiceFromAddress(PaymentMethodId paymentMe
...
+                var owner = await db.ExecuteScalarAsync<string>(
+                    """
+                    SELECT "InvoiceDataId" FROM "AddressInvoices"
+                    WHERE "Address" = @address AND "PaymentMethodId" = @paymentMethodId
+                    """,
+                    new { address, paymentMethodId },
+                    context.Database.CurrentTransaction?.GetDbTransaction());
+                if (owner is not null)
+                    return owner;
+            }
```
[the remark this answers is not in the document]
Comment (CaMoPeZzz) [reply BTCPayServer/Services/Invoices/InvoiceRepository.cs]: Fixed in 84389baea. The owner lookup retry is bounded, has incremental backoff, and propagates cancellation through both Dapper commands and the delay.

File: BTCPayServer/Services/Invoices/InvoiceRepository.cs [hunk 5]
```diff
@@ -346,77 +405,122 @@ public Task UpdatePrompt(string invoiceId, PaymentPrompt prompt)
...
+                            }
+                            var search = new List<String>();
+                            search.Add(prompt.Calculate().TotalDue.ToString());
+                            invoiceEntity.SetPaymentPrompt(prompt.PaymentMethodId, prompt);
+                            invoice.SetBlob(invoiceEntity);
+                            AddToTextSearch(db, invoice, search.Concat(paymentPromptContext.AdditionalSearchTerms).Distinct().ToArray());
+                            await db.SaveChangesAsync(cancellationToken);
+                            return true;
+                        },
+                        (_, cancellationToken) => IsPromptPersisted(invoiceId, prompt, destinations, cancellationToken));
```
[the remark this answers is not in the document]
Comment (CaMoPeZzz) [reply BTCPayServer/Services/Invoices/InvoiceRepository.cs]: Fixed in 84389baea. The verifier now receives an empty destination set whenever the prompt is inactive or has no destination, matching the claim operation.

**Comment by CaMoPeZzz** [comment]
Addressed all review findings in 84389baea, including the test helper cleanup.

Validation: all 12 `DatabaseTests` pass locally on .NET SDK 10.0.302.

**Comment by NicolasDorier** [comment]
Concept ACK, but I can't review this code at this time it is too confusing.
Like why does it even touch lightning listener?

Also I'd prefer a simpler test:
* Generate an invoice with Lightning and BTC
* "Unreserve" the generated address on NBXplorer, so next time it will generate the same address.
* Generate a new invoice.
* Verify only lightning is present
* Check the logs says that there is a dups.

That way we can better confirm it does what it says it does.

**Comment by CaMoPeZzz** [comment]
Yes, makes sense. I went too far and tried to fix a bunch of related edge cases in one go (It's all on the AI review 🙃).

I’ve cut the scope back to invoice creation only. No more changes in LightningListener.
I also redid the tests, it turned out much easier, thx

**Comment by NicolasDorier** [comment]
Pushed commit, several fixes were needed:
1. Avoiding 5 DB roundtrip if there was 5 payment methods
2. Make sure this fix also apply to invoices with LazyPaymentMethod

**Comment by CaMoPeZzz** [comment]
@NicolasDorier anything else to fix?

**Comment by napoly** [comment]
@CaMoPeZzz you mentioned Monero as an example.. but how would that work in a real wold scenario..? the customer seeking privacy.. is redirected and risking to be jailed ?

**Comment by CaMoPeZzz** [comment]
> @CaMoPeZzz you mentioned Monero as an example.. but how would that work in a real wold scenario..? the customer seeking privacy.. is redirected and risking to be jailed ?

This fix doesn’t add any redirects or change the customer flow. It only prevents invoice creation from failing when a plugin returns a destination already used by another invoice.

**Comment by NicolasDorier** [comment]
@CaMoPeZzz rebase it on master, so the tests can run.

**Comment by CaMoPeZzz** [comment]
@NicolasDorier pls approve
изображение

**Comment by rockstardev** [comment]
@CaMoPeZzz sorry for delay, running now and will bump @NicolasDorier on this one
